src/Tools/dotnet-openapi/src/Commands/BaseCommand.cs

```diff
@@ -0,0 +1,308 @@
+            }
+            else
+            {
+                var projects = Directory.GetFiles(WorkingDirectory, "*.csproj", SearchOption.TopDirectoryOnly);
+                if (projects.Length == 0)
+                {
+                    Error.Write("No project files were found in the current directory. Either move to a new directory or provide the project explicitly");
+                }
+                if (projects.Length > 1)
+                {
+                    Error.Write("More than one project was found in this directory, either remove a duplicate or explicitly provide the project.");
+                }
+
+                project = projects[0];
+            }
+
+            return new FileInfo(project);
+        }
+
+        protected Project LoadProject(FileInfo projectFile)
+        {
+            var project = ProjectCollection.GlobalProjectCollection.LoadProject(
+                projectFile.FullName,
+                globalProperties: null,
+                toolsVersion: null);
+            project.ReevaluateIfNecessary();
+            return project;
+        }
+
+        internal bool IsProjectFile(string file)
+        {
+            return File.Exists(Path.GetFullPath(file)) && file.EndsWith(".csproj");
+        }
+
+        internal bool IsUrl(string file)
+        {
+            return Uri.TryCreate(file, UriKind.Absolute, out var _) && file.StartsWith("http");
+        }
+
+        internal void AddServiceReference(
+            string tagName,
+            FileInfo projectFile,
+            string sourceFile,
+            string sourceUrl = null)
+        {
+            var project = LoadProject(projectFile);
+            var items = project.GetItems(tagName);
+            var item = items.FirstOrDefault((i) => {
+                return string.Equals(GetFullPath(i.EvaluatedInclude), GetFullPath(sourceFile), StringComparison.Ordinal);
+            });
+
+            if (sourceUrl != null)
+            {
+                var urlMatch = items.SingleOrDefault(i => string.Equals(i.GetMetadataValue(SourceUrlAttrName), sourceUrl));
+                if (urlMatch != null)
+                {
+                    Out.Write($"A reference to '{sourceUrl}' already exists in '{project.FullPath}'.");
+                    return;
+                }
+            }
+
+            if (item == null)
+            {
+                var metadata = new Dictionary<string, string>();
+
+                if (!string.IsNullOrEmpty(sourceUrl))
+                {
+                    metadata[SourceUrlAttrName] = sourceUrl;
+                }
+
+                project.AddElementWithAttributes(tagName, sourceFile, metadata);
+            }
+            else
+            {
+                Out.Write($"A reference to '{sourceFile}' already exists in '{project.FullPath}'.");
+            }
+        }
+
+        internal async Task DownloadToFileAsync(string url, string destinationPath, bool overwrite)
+        {
+            Application application;
```

Comment:
This is a weird pattern... same when it's used for working directory.

Reply:
Do you ever change the working Directory or need a different DownloadProvider (also not sure why the Provider suffix, should be something like DownloadAsync)? If not, just resolve working dir when needed and use a static DownloadAsync method on the base class here.

Reply:
DownloadProvider is given a different value in the tests so that we're not reliant on an http request as per https://github.com/aspnet/AspNetCore/pull/10015#discussion_r281457154.